// ngx-file-drop/esm2015/ngx-file-drop/ngx-file-drop.module.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes,extraRequire,missingOverride,missingReturn,unusedPrivateMembers,uselessCode} checked by tsc
 */
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { NgxFileDropComponent } from './ngx-file-drop.component';
import { NgxFileDropContentTemplateDirective } from './ngx-templates.directive';
export class NgxFileDropModule {
}
NgxFileDropModule.decorators = [
    { type: NgModule, args: [{
                declarations: [
                    NgxFileDropComponent,
                    NgxFileDropContentTemplateDirective,
                ],
                imports: [
                    CommonModule
                ],
                exports: [
                    NgxFileDropComponent,
                    NgxFileDropContentTemplateDirective,
                ],
                providers: [],
                bootstrap: [
                    NgxFileDropComponent
                ],
            },] }
];
